sw/device/lib/crypto/aes_gcm/aes_gcm.c

```diff
@@ -354,3 +372,97 @@ aes_error_t aes_gcm_ghash(const aes_block_t *hash_subkey,
 
   return kAesOk;
 }
+
+/**
+ * One-shot version of the AES API for a single block, for convenience.
+ */
+static aes_error_t aes_single_block(const aes_params_t params,
+                                    const aes_block_t *input,
+                                    aes_block_t *output) {
+  aes_error_t err = aes_begin(params);
+  if (err != kAesOk) {
+    return err;
+  }
+  err = aes_update(output, input);
+  if (err != kAesOk) {
+    return err;
+  }
+  err = aes_end();
+  if (err != kAesOk) {
+    return err;
+  }
+  return kAesOk;
+}
+
+/**
+ * Implements the GCTR function as specified in SP800-38D, section 6.5.
+ *
+ * The block cipher is fixed to AES. Note that the GCTR function is a modified
+ * version of the AES-CTR mode of encryption.
+ *
+ * Input must be less than 2^32 blocks long; that is, input_len < 2^28, since
+ * each block is 16 bytes.
+ *
+ * @param aes_ctr_params AES parameters for CTR encryption with IV set to the
+ * initial counter block
+ * @param input_len Number of bytes in the input
+ * @param input Pointer to input buffer (may be NULL if input_len is 0)
+ * @param output Pointer to output buffer (same size as input, may be the same
+ * buffer)
+ */
+aes_error_t aes_gcm_gctr(aes_params_t aes_ctr_params, const size_t input_len,
+                         const uint8_t *input, uint8_t *output) {
+  // If the input is empty, the output must be as well. Since the output length
+  // is 0, simply return.
+  if (input_len == 0) {
+    return kAesOk;
+  }
+
+  // NULL pointers are not allowed for nonzero input length.
+  if (input == NULL || output == NULL) {
+    return kAesInternalError;
+  }
+
+  // Check the input parameters.
+  if (aes_ctr_params.mode != kAesCipherModeCtr || !aes_ctr_params.encrypt) {
```

Comment:
Are you planning to have a wrapper for this function setting the mode? Otherwise I'd set the mode here.